core/src/main/java/org/apache/druid/timeline/partition/OvershadowableManager.java

```diff
@@ -105,12 +105,46 @@
     this.overshadowedGroups = new TreeMap<>();
   }
 
-  OvershadowableManager(OvershadowableManager<T> other)
+  public OvershadowableManager<T> copyVisible()
   {
-    this.knownPartitionChunks = new HashMap<>(other.knownPartitionChunks);
-    this.standbyGroups = new TreeMap<>(other.standbyGroups);
-    this.visibleGroupPerRange = new TreeMap<>(other.visibleGroupPerRange);
-    this.overshadowedGroups = new TreeMap<>(other.overshadowedGroups);
+    final OvershadowableManager<T> copy = new OvershadowableManager<>();
+    visibleGroupPerRange.forEach((partitionRange, versionToGroups) -> {
+      // There should be only one group per partition range
+      final AtomicUpdateGroup<T> group = versionToGroups.values().iterator().next();
+      group.getChunks().forEach(chunk -> copy.knownPartitionChunks.put(chunk.getChunkNumber(), chunk));
+
+      copy.visibleGroupPerRange.put(
+          partitionRange,
+          new SingleEntryShort2ObjectSortedMap<>(group.getMinorVersion(), AtomicUpdateGroup.copy(group))
+      );
+    });
+    return copy;
+  }
+
+  public OvershadowableManager<T> deepCopy()
+  {
+    final OvershadowableManager<T> copy = copyVisible();
+    overshadowedGroups.forEach((partitionRange, versionToGroups) -> {
+      // There should be only one group per partition range
+      final AtomicUpdateGroup<T> group = versionToGroups.values().iterator().next();
+      group.getChunks().forEach(chunk -> copy.knownPartitionChunks.put(chunk.getChunkNumber(), chunk));
+
+      copy.overshadowedGroups.put(
+          partitionRange,
+          new SingleEntryShort2ObjectSortedMap<>(group.getMinorVersion(), AtomicUpdateGroup.copy(group))
+      );
+    });
+    standbyGroups.forEach((partitionRange, versionToGroups) -> {
+      // There should be only one group per partition range
+      final AtomicUpdateGroup<T> group = versionToGroups.values().iterator().next();
+      group.getChunks().forEach(chunk -> copy.knownPartitionChunks.put(chunk.getChunkNumber(), chunk));
+
+      copy.standbyGroups.put(
+          partitionRange,
+          new SingleEntryShort2ObjectSortedMap<>(group.getMinorVersion(), AtomicUpdateGroup.copy(group))
+      );
+    });
+    return copy;
```

Comment:
missing tests to verify that the returned copy is a `deepCopy()` - what if new fields are added to the overshadowable manager? Similar comments for `copyVisible()`

Reply:
Added tests.